### PR TITLE
bpo-29659: Expose copyfileobj() `length` arg for public use

### PR DESCRIPTION
**Pull request by goodboy**
Corresponds to issue bpo-29659

When copying large files the copy time can be sufficiently
shortened by increasing the memory buffer used in the `copyfileobj()`
routine.  Expose the `length` argument from `copyfileobj()` upwards
for use throughout the module.

https://bugs.python.org/issue29659

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow these steps to rectify the issue:

1. Sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username" requested by the form is the "Login name" field in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
2. **Wait at least one US business day** and then check the "Contributor form received entry under "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
3. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

**Comment by goodboy** [comment]
Hello,

I have completed the above steps :)

Thanks!

Tyler Goodlet

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: I suggest:

if not length:
   length = 16 * 1024

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: You should document the default size: 16 kB.

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: Same: document default size.

**Review by vstinner — CHANGES_REQUESTED** [review]
You must update Doc/library/shutil.rst: document the new parameter, but also add a ".. versionchanged:: 3.7" section to document the change.

File: Lib/shutil.py
Comment (goodboy) [reply Lib/shutil.py]: Will do :+1:

File: Lib/shutil.py
Comment (goodboy) [reply Lib/shutil.py]: Sounds good!

**Comment by goodboy** [comment]
@haypo I believe I've addressed everything!
If there's anything I'm missing please let me know :)

File: Doc/library/shutil.rst
Comment (vstinner) [inline Doc/library/shutil.rst]: I wouldn'd say "expose ..." but just "Added length parameter", and versionadded is not the right place to document "default is ...", put it in the documentation body.

You didn't update copyfile() signature. Same remark for the 2 other functions.

Moreover, you should also document the length in the function document, not only in versionadded. The "documentation" can just ask to refer to copyfileobj().

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: please mention the unit, "size in bytes"

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/shutil.py
Comment (goodboy) [reply Lib/shutil.py]: @haypo to clarify, do you prefer:
`'An in-memory buffer size can be set with `length`; the default size in bytes is 16*1024'`, 
or 
`"An in-memory buffer size in bytes can be set with `length`; the default is 16 KB.`
?

File: Doc/library/shutil.rst
Comment (goodboy) [reply Doc/library/shutil.rst]: Shoot yeah not sure how I missed the signatures...

File: Lib/shutil.py
Comment (vstinner) [reply Lib/shutil.py]: "An in-memory buffer size in bytes can be set withlength; the default is 16 KB."

**Comment by goodboy** [comment]
@haypo hopefully 3rd times the charm ;)

If you'd like me to squash the history to one commit let me know!

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Suggestion: leave API of copyfileobj should alone as 'length=16*1024' to document default.  For other functions, say something like 'If length not passed, copyfileobj uses its default.'  We might change it someday.

Required: new or revised tests that use the expanded api (and fail without patch).

**Comment by goodboy** [comment]
@terryjreedy the only problem with that is that we'll need to have the `if length` checks in all consuming functions as opposed to only once in `copyfileobj` no? Maybe you have a better approach?

Yes I totally agree about the test. I guess it goes [here?](https://github.com/python/cpython/blob/master/Lib/test/test_shutil.py#L1900)

**Comment by vstinner** [comment]
> Yes I totally agree about the test. I guess it goes here?
<https://github.com/python/cpython/blob/master/Lib/test/test_shutil.py#L1900>

Yes.

File: Doc/library/shutil.rst
Comment (berkerpeksag) [inline Doc/library/shutil.rst]: Indentation should use three spaces here:

```rst
.. versionchanged:: 3.7
   Added the *length* parameter.
```

Comment (berkerpeksag) [reply Doc/library/shutil.rst]: \`length\` -> \*length\*

(and please finish the sentence with a full stop.)

Comment (goodboy) [reply Doc/library/shutil.rst]: Done.

File: Lib/shutil.py
Comment (vadmium) [reply Lib/shutil.py]: I encourage you to use _KiB_, which other parts of the documentation already use. If you write _16 KB_ it could easily be interpreted as 16,000 B (lower-case _k_ means 1000, as in _km_ for kilometre).

File: Doc/library/shutil.rst
Comment (vadmium) [inline Doc/library/shutil.rst]: IMO “Note that” is redundant. In fact the whole first line could be dropped, and “Only the contents from the current file position . . .” could be put back in the first paragraph.

Comment (terryjreedy) [reply Doc/library/shutil.rst]: I agree with this.

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: ditto: KiB

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: PEP 8: add spaces around operators, "16 * 1024"

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py]: Please remove this empty line.

**Review by vstinner — COMMENTED** [review]
LGTM, but please address pending comments before I can approve the change ;-)

**Comment by vstinner** [comment]
Note: I created PR #4293 to replace KB with KiB in the Python code base ;-)

File: Doc/library/shutil.rst
Comment (goodboy) [reply Doc/library/shutil.rst]: Done!

File: Lib/shutil.py
Comment (goodboy) [reply Lib/shutil.py]: done

File: Lib/shutil.py
Comment (goodboy) [reply Lib/shutil.py]: done

File: Lib/shutil.py
Comment (goodboy) [reply Lib/shutil.py]: done

**Comment by goodboy** [comment]
I think everything has been addressed except for @terryjreedy's:
>Suggestion: leave API of copyfileobj should alone as 'length=16*1024' to document default. For other functions, say something like 'If length not passed, copyfileobj uses its default.' We might change it someday.

I don't know how to keep the default value in `copyfileobj` but also support it as a default argument in all client functions without `length = 16 * 1024 if length is None else length` in each case. Personally I think that's a bit uglier. I can also stick the default value in each client function to be `length=16 * 1024` and always pass it through transparently if that is preferred; it just means changing the value in multiple places if it is changed in the future. I had originally opted for one change in one place if it was to be modified later.

Whatever you all think is best I'll do 👍

File: Lib/test/test_shutil.py
Comment (vstinner) [inline Lib/test/test_shutil.py]: I suggest to use a smaller file: write('x' * 100) and then length=20.

By the way, I suggest to use tempfile.NamedTemporaryFile rather than temporary files:

```
            with tempfile.NamedTemporaryFile() as src:
                with tempfile.NamedTemporaryFile() as dst:
                    write_file(src.name, b'x' * 100, binary=True)
                    fn(src.name, dst.name, length=20)
```

Comment (goodboy) [reply Lib/test/test_shutil.py]: Hehe thanks @vstinner! You did the work for me 👍

File: Lib/shutil.py
Comment (AraHaan) [inline Lib/shutil.py:76]:
```py
def copyfileobj(fsrc, fdst, length=16*1024):
```
I personally prefer this way in my code to concentrate 2 lines worth of code into one.

```py
    if not length:
        length = 16 * 1024
```
It looks a little nicer and not to mention saves lines of code that basically do the same thing.
And yes I am also guilty of doing it this way and then people started hating me for it and calling my a bad programmer for it.

**Comment by goodboy** [comment]
Just to clear this up since it seems my original reasoning isn't showing well and the drive-by comments are piling up.

Afaict there are 3 ways to deal with the default value of `length`:

- stick `length=16*1024` in all three function signatures for `copyfileobj`, `copy`, and `copy2`
  * this means any time the default needs to be changed it will be done in 3 places
  * this is the least amount of lines of code

- keep it the way it is (i.e. `copy` and `copy2` define `length=None` in their signatures and one function, `copyfileobj`, has the current `if length is None:` check)
  * one more line of code then the previous solution
   * have to change the `length` in one place in the future
   * `length` is an explicit default argument in `copy` and `copy2`

- make `copy` and `copy2` take `**kwargs` and pass them through to the internal call to `copyfileobj`
  * same lines of code as the first solution
  * doesn't explicitly document `length` in each of the the `copy`, `copy2` signatures
  * one place to change the default value in the future

If one of these solutions is preferred over the others after acknowledging the trade-offs I'm fine to do it, but just simply saying `length=16*1024` is less code doesn't deal with how to handle `copy` and `copy2` and the extra (or less clear) code they may contain.

**Comment by goodboy** [comment]
@vstinner looks like the appveyor build failed due to the `with open(path, 'wb' if binary else 'w') as fp:` line?

**Comment by vstinner** [comment]
> @vstinner looks like the appveyor build failed due to the with open(path, 'wb' if binary else 'w') as fp: line?

```
ERROR: test_copy_w_different_length (test.test_shutil.TestShutil)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\cpython\lib\test\test_shutil.py", line 1417, in test_copy_w_different_length
    write_file(src.name, b'x' * 100, binary=True)
  File "C:\projects\cpython\lib\test\test_shutil.py", line 60, in write_file
    with open(path, 'wb' if binary else 'w') as fp:
PermissionError: [Errno 13] Permission denied: 'C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\tmpj1vhtx1b'
```

Sorry, I don't know why the test failed.

**Comment by AraHaan** [comment]
Looks to me like an normal permission error as python seems to not have permissions to write to the temp folder? Maybe somehow have python write to somewhere that does not require administrator permissions? It would be nice if python had an nice function that you provide the path and it checks if you have permissions to write to there to avoid having to try / except permission errors. 🤔 That function would be the perfect thing to fix this test on Windows.

**Comment by tgoodlet** [comment]
Ok so besides the outstanding:
- missing news entry
- failing test

Is there anything else in the patch that needs to be addressed?

**Comment by AraHaan** [comment]
I dont think so.

File: Lib/shutil.py
Comment (vstinner) [inline Lib/shutil.py:84]: I'm not sure that it's ok to use a loop if the length is negative. I suggest to have a special case for negative value calling read() (no parameter) only *once*.

File: Doc/library/shutil.rst
Comment (vstinner) [inline Doc/library/shutil.rst:49]: "A negative *length* value means to copy the data without looping over the source data in chunks"

I dislike this definition. In practice, negative means "unlimited" buffer size: the whole input file is loaded into memory.

I'm not sure that it's a good practice to try to load files of unknown size into memory.

I suggest to remove this feature which seems more like a side effect than a carefully designed API.

If you want to get fast copy, pass a very large length like 1 GB. But if Python starts to load 1 TB into memory, it's likely to crash the system... At least, to slow down the system, a lot.

**Review by vstinner — CHANGES_REQUESTED** [review]
I dislike length=-1 "feature".

**Comment by vstinner** [comment]
Fixed by https://bugs.python.org/issue33671

**Pull request closed without merging**
